serialization/ascii/unmarshal_utils.go

```diff
@@ -0,0 +1,179 @@
+package ascii
+
+import (
+	"fmt"
+	"reflect"
+)
+
+func errInvalidData(p []byte) error {
+	for i := range p {
+		if p[i] > 127 {
+			return fmt.Errorf("failed to unmarshal ascii: invalid charester %s", string(p[i]))
+		}
+	}
+	return nil
+}
+
+func errNilReference(v interface{}) error {
+	return fmt.Errorf("failed to unmarshal ascii: can not unmarshal into nil reference(%T)(%[1]v)", v)
+}
+
+func DecString(p []byte, v *string) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	*v = decString(p)
+	return errInvalidData(p)
+}
+
+func DecStringR(p []byte, v **string) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	*v = decStringR(p)
+	return errInvalidData(p)
+}
+
+func DecBytes(p []byte, v *[]byte) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	*v = decBytes(p)
+	return errInvalidData(p)
+}
+
+func DecBytesR(p []byte, v **[]byte) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	*v = decBytesR(p)
+	return errInvalidData(p)
+}
+
+func DecInterface(p []byte, v *interface{}) error {
+	if v == nil {
+		return errNilReference(v)
+	}
+	*v = decString(p)
+	return errInvalidData(p)
+}
+
+func DecReflect(p []byte, v reflect.Value) error {
+	if v.IsNil() {
+		return errNilReference(v)
+	}
+
+	switch v = v.Elem(); v.Kind() {
+	case reflect.String:
+		v.SetString(decString(p))
+	case reflect.Slice:
+		if v.Type().Elem().Kind() != reflect.Uint8 {
+			return fmt.Errorf("failed to marshal ascii: unsupported value type (%T)(%[1]v)", v.Interface())
+		}
+		v.SetBytes(decBytes(p))
+	case reflect.Interface:
+		v.Set(reflect.ValueOf(decString(p)))
+	default:
+		return fmt.Errorf("failed to unmarshal ascii: unsupported value type (%T)(%[1]v)", v.Interface())
+	}
+	return errInvalidData(p)
+}
+
+func DecReflectR(p []byte, v reflect.Value) error {
+	if v.IsNil() {
+		return errNilReference(v)
+	}
+
+	switch ev := v.Type().Elem().Elem(); ev.Kind() {
+	case reflect.String:
+		return decReflectStringR(p, v)
+	case reflect.Slice:
+		if ev.Elem().Kind() != reflect.Uint8 {
+			return fmt.Errorf("failed to marshal ascii: unsupported value type (%T)(%[1]v)", v.Interface())
+		}
+		return decReflectBytesR(p, v)
+	default:
+		return fmt.Errorf("failed to unmarshal ascii: unsupported value type (%T)(%[1]v)", v.Interface())
+	}
+}
+
+func decReflectNullableR(p []byte, v reflect.Value) reflect.Value {
```

Comment:
I see it is used only once, please inline it into the place it is used

Reply:
Fixed.
Also fixed conversations from PR #321.

Reply:
Thanks.